### DISCUSSION TITLE
Relation to self not eagerly populating past 2 levels of nesting

### DISCUSSION THREAD

**Discussion by gvonhagel** [mikro-orm/mikro-orm#3354] [Q&A]
**Describe the bug**

I have an entity that has a `@OneToMany` relationship with itself, one parent can have many children, can have many children, etc.

```
@Entity()
export class Group extends BaseEntity<Group, 'id'> {
  @PrimaryKey()
  id!: number;

  @Property()
  name: string;

  @ManyToOne(() => Group, { nullable: true })
  parent: Group;

  @OneToMany(() => Group, (group) => group.parent, { eager: true })
  subGroups = new Collection<Group>(this);

}
```

When I have a table that looks something like:

```
+--+---------+----------------+
|id|parent_id|name            |
+--+---------+----------------+
|1 |null     |Parent Group    |
|2 |1        |SubGroup Level 1|
|3 |2        |SubGroup Level 2|
|4 |3        |SubGroup Level 3|
|5 |4        |SubGroup Level 4|
+--+---------+----------------+
```

And call `GroupRepository.find()` I get `subGroup` Collections up to 3 levels deep. The queries are still being made in the background as seen here:

```
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."id" = 1 limit 1 [took 0 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (1) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (2) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (3) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (4) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (5) order by "a0"."parent_id" asc [took 1 ms]
```

But the mapped object returned only goes 3 levels deep:

```
{
  "id": 1,
  "name": "Parent Group",
  "parent": null,
  "subGroups": [
    {
      "id": 2,
      "name": "SubGroup Level 1",
      "parent": 1,
      "subGroups": [
        {
          "id": 3,
          "name": "SubGroup Level 2",
          "parent": 2
        }
      ]
    }
  ]
}
```

**Expected behavior**
The query should return `subGroups` nested Collections as long as more rows are returned matching `parentId`. If this is the intended behavior queries should not continue to be made without their results being appended to the final object.

**Versions**

| Dependency | Version |
| - | - |
| node | 16.16.0 |
| typescript | 4.7.4 |
| mikro-orm | 5.2.4 |
| pg | 8.7.3 |

**Comment by B4nan** [comment]
Dang, I had quite a long response prepared for this but managed to delete it :D 

So in a nutshell - it works correct, what you see is a serialization issue, the entity (on runtime) has everything propertly populated. The problem here is using `eager: true`, you need to use populate hints to let the ORM know how things should be serialized.

Changing this behaviour would be breaking change I'm afraid.

You can get around this either by providing the populate hint:

```ts
const res = await orm.em.find(Group, {}, { populate: ['subGroups.subGroups.subGroups.subGroups'] });
console.log(JSON.stringify(res));
```

Or by using `toPOJO` which ignores populate hints:

```ts
const res = await orm.em.find(Group, {});
const jsons = res.map(g => wrap(g).toPOJO())
console.log(JSON.stringify(jsons));
```

**Reply by gvonhagel** [reply]
Thanks for the quick response! If the number of `subGroups` is indeterminate, I assume the second solution is the only viable one? There's no shorthand in `populate` like `subGroups.*` that would work here, correct?

**Reply by gvonhagel** [reply]
Seems like I have to recursively call `toPOJO()` on each nested level of `subGroups` here. Calling `toPOJO` on the top-level parent is not sufficient

**Reply by B4nan** [reply]
It should be, it works fine in my tests based on your code. It will already use itself recursively.

**Reply by gvonhagel** [reply]
Using the following code:
```
  async findOne(requester: User, id: number): Promise<Group> {
    const group = await this.entityManager.findOne(Group, { id });
    const output: any = wrap(group).toPOJO(); // Group extends BaseEntity, so this could be unwrapped 
    console.log(JSON.stringify(output));
    return output;
  }
  ```
  
  The following queries are executed:
```
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."id" = 1 limit 1 [took 0 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (1) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (2) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (3) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (4) order by "a0"."parent_id" asc [took 1 ms]
[query] select "a0"."id", "a0"."name", "a0"."parent_id" from "asset_group" as "a0" where "a0"."parent_id" in (5) order by "a0"."parent_id" asc [took 1 ms]
```

Only returns a `Group` with one nested `subGroup` - from the console.log:

```
{
  "id": 1,
  "name": "Parent Group",
  "parent": null,
  "subGroups": [
    {
      "id": 2,
      "name": "SubGroup 1",
      "parent": {
        "id": 1,
        "name": "Parent Group"
      }
    }
  ]
}
```

**Interestingly,** if I modify the function to look like this:

```
  async findOne(requester: User, id: number): Promise<Group> {
    const group = await this.entityManager.findOne(Group, { id });
    const output: any = wrap(group).toPOJO(); // Group extends BaseEntity, so this could be unwrapped 
    console.log(JSON.stringify(output));
    return  wrap(group).toPOJO();
  }
  ```

then I get one additional level of nesting in the _return type_ (not in the original `console.log`).

Am I missing something here? Now I'm properly confused.

**Reply by B4nan** [reply]
I am sure this is just bad observation, this cant have any difference.

edit: hmm this is really weird, I can see the same locally too, and if dont call `toPOJO` twice it really does not work properly

**Reply by gvonhagel** [reply]
Maybe should change this back into a bug ticket? :sweat_smile:

**Reply by B4nan** [reply]
The fact that it works on the second run is actually a bug, the serialization context is not cleared properly (got that fixed locally already). Even with `toPOJO` there is a cycle detection that is triggered, will need to give it more time to see if we can do something about it.

**Reply by B4nan** [reply]
Here you go: https://github.com/mikro-orm/mikro-orm/commit/aa10802e53e0d6500a02230b01292de08f9fb9bc
